proposals/module-types/Explainer.md

```diff
@@ -0,0 +1,745 @@
+`$identifier`. (Also, just like function types, the inline form is an
+[abbreviation][typeuse-abbrev] for an implicit type definition and use.) For
+example, an instance type can be defined:
+```wasm
+(type $WasiFile (instance
+  (export "read" (func (param i32 i32 i32) (result i32)))
+  (export "write" (func (param i32 i32 i32) (result i32)))
+))
+```
+
+In many examples shown below, type definitions are needed for *both* a module
+type and the instance type produced when that module type is instantiated. In
+such cases, to avoid duplicating all the exports, a new `(exports $instanceType)`
+type expression is added which can be used to define module types. For example,
+here is the type of a module which implements the above-defined `$WasiFile`
+interface via Win32 operations:
+```wasm
+(module
+  (import "Win32" "ReadFile" (func (param i32 i32 i32 i32) (result i32)))
+  (import "Win32" "WriteFile" (func (param i32 i32 i32 i32) (result i32)))
+  (exports $WasiFile)
+)
+```
+
+Lastly, just as the current [text format conventions recommend `.wat`][WAT] as
+the extension of a file that contains a module definition, this proposal
+includes a new recommendation for text files containing a bare module or
+instance type be suffixed `.wit`. `.wit` files can be used as part of the
+toolchain ecosystem for describing a module's interface without including its
+definition. For example, this can be used to generate compatible source-language
+declarations. Like `.wat` files, `.wit` files are not consumed directly by the
+host.
+
+Module and Instance types can also be used to describe the types of imports:
+
+
+### Instance Imports
+
+Just as a function, memory, table or global can be imported by specifying a
+function, memory, table or global type, modules or instances can be imported
+by specifying a module or instance type.
+
+A single instance import is practically equivalent to individually importing
+all of its exports. Indeed, instances have no distinct identity: instances are
+just [immutable tuples][`moduleinst`] containing the *addresses* of mutable
+things with identity (that could be independently imported) like memories and
+tables. For example, the instance import in this module:
+```wasm
+(module
+  (import "i" (instance $i
+    (export "f1" (func $f1))
+    (export "f2" (func $f2 (param i32)))
+  ))
+  (func (export "run")
+    call $i.$f1
+  )
+)
+```
+is practically equivalent to individually importing the instance's exports:
+```wasm
+(module
+  (import "i" "f1" (func $f1))
+  (import "i" "f2" (func $f2 (param i32)))
+  (func (export "run")
+    call $f1
+  )
+)
+```
+In particular, the exports of an instance import are injected into the
+appropriate [index spaces] of the containing instance, allowing them to be
+directly accessed by `call`, `i32.load`, `table.get`, etc. The `$i.$f` textual
+notation is new, but still encodes in the binary format as a single integer
```

Comment:
Maybe it could be kept mostly-syntactic by saying that, when an instance import or definition is parsed, all its dotted identifiers are eagerly injected (as single strings `$a.$b`) into the identifier scope, such that lookup is done blindly regardless of `.`?

Reply:
Yes,  but since the knowledge of what's injected is not local to the definition (it depends on a separate module definition, whose scope in turn can depend on type definitions, and so on). So that still requires maintaining all sorts of direct and indirect type information.

FWIW, you'll have similar problems wrt to indexing in a decoder as well. For example, the size of an index space no longer corresponds to the size of the associated section (plus imports), since members of all instance declarations get spliced into it. And to know what's in there you'll need to look at modules and their types. So it becomes very difficult to determine the size of index spaces and resp tables upfront, which I think would undermine the single pass / streaming design goal.

(Worse, once you consider nested instances I think it will be necessary to support `$i.$j.$f` and longer paths if you want to be able to access inner instances. But then the whole _tree_ of indexes has to get flattened into the outer index space.)

The more I think about all this, the nastier it sounds to me.

Here's what I suggest instead: Extend the notion of aliases and treat dot names as text format sugar for that.

That is, uniformly allow defining aliases for all sorts of entities. Probably as a new alias section that also adds to all index spaces, similar to the import section. The RHS of an alias is either an outer scope index (for types and modules) or a single projection from an inner scope, i.e., an instance in the current scope.

Once we have that, you can treat `$x.$y` as textual sugar for an inline alias definition (or several, for longer paths) that implicitly gets added to the module, similar to what we do with inline function types. The resolved AST continues to only deal with flat index spaces. WDYT?

Reply:
Ooh yeah, I like that.  I had been sortof uncomfortable with the fact that indices were indiscriminately injected which in some sense takes away a module's control over its own index space.

Reply:
Trying to write up how the generalized aliases scheme would work in more detail, I don't see how we can get away with something as simple as a new "aliases" section: you end up needing to put various aliases in various places scattered throughout the module:
* in the type section as currently proposed
* after the import section, before the first instance definition to make imported instances' exports available
* between any two instance definitions to make the previous instance's exports available to the next

So I wonder if the automatic injection is indeed simpler.  Regarding my "control over index space" concern: an outer module does have *some* control because it can always narrow down the set of injected exports (due to module subtyping).  Regarding your steaming concern, I don't think there actually is one: fundamentally what you need for parallel+streaming compilation is an immutable environment under which to compile the *code* section.  (Code sections (+ inert data/elem sections) make up >99% of the bytes.)  If we assume the code section of a module `M` comes after all the nested instances of `M`, that would be satisfied.  Moreover, it would also be true for every code section of every nested module of `M`.

Reply:
Who says that there must only be one alias section? ;) I think allowing multiple ones is still way simpler and cleaner than index space flattening.

I might be even worse than I thought at first. I only now realised that flattening nested index spaces is similar to inserting an `import *` or ML-style `open`. And that is known to be hairy when scoping isn't strictly sequential. When I looked into early ideas for the JS module system with Dave & Sam TH a long time ago, we actually tried to specify a design with `import *`, but could not define it in a way that wouldn't require a fixpoint iteration for the scope analysis. For ML, there even is a result that (the somewhat more general problem of) dependency analysis in the presence of `open` is NP-complete. Roughly, the problem is that resolving inner and outer scopes becomes mutually dependent -- you cannot tell what an inner scope contains before knowing what a reference to an outer scope resolves to, and you cannot know that before you have resolved enough inner scopes. IIRC, there might even be cases where there is more than one solution.

I don't currently have a good feeling for how much of that hardness would carry over here, but I would bet that it turns out to be complicated.

As for streaming, nested modules now contain nested code sections. That would break the 99% assumption, unless nested modules are placed after the code section? Are you assuming that modules need to forward-declare their types in a separate section, like functions? Unlike for functions, that would require substantial extra checking (checking their actual against their declared signatures), and potentially duplicating type definitions.

Reply:
Ok, I think there's actually a minimal tweak to the existing proposal that doesn't involve significantly changing the way sections work today:
* keep the existing design where the definitions in the type section and module section can alias parent types/modules via a new `alias` case
* in the instance section, add a new `alias` case that explicitly injects a single export of any preceding instance in the instance index space (which could be an instance, thereby allowing paths like `$i.$j.$k` via a sequence of aliases).

Really, this is isomorphic to allowing type/module/instance sections to be split up and interleaved with alias sections, just simpler and more compact.

FWIW, I don't think the `import *` issues you've seen in JS/ML would apply to wasm because of our "strong modularity" property: I can know all I need to know about a nested modules' exports just by looking at that module in isolation.  Moreover, I think one would be able to express all cases of flattening with `alias`es; therefore flattening should be no less stream-compilable.

But I do find the explicit index-space control vaguely nicer; e.g., one concrete benefit is that you'd never "accidentally" pull in a memory that takes the coveted role of default memory (memory index 0), so I'll switch.

Reply:
Ok, pushed the change, PTAL.  (I'll resolve this conversation to keep the size manageable; feel free to start new ones.)

Reply:
(Reopening this one since I'm replying to the comment.)

That may work, but won't it still be necessary to interleave this with other sections? For example, when a module declaration wants to refer to a type from an imported instance? Moreover, isn't it necessary to be able to interleave module and (actual) instance declarations? If a module declaration needs to refer to types produced by an instance then it's not possible to declare all modules before all instances and their projections.

(FTR, the `import *` problem is present as soon as an inner scope can refer to definitions from an outer, which IIUC, you are proposing wrt types and modules, which in turn implies the ability to (indirectly) refer to instances.)

Reply:
Hrm, if we want to support all of *that*, it sounds like really then we need the ability to arbitrarily interleave module, instance *and* alias definitions.  That suggests having a "Nested" section, containing all 3 definitions, right after the Import Section.

Reply:
Ok, the most recent commit should address this.  Feel free to reopen or start new convo on new text.